src/mnode/mnode-pastel.cpp

```diff
@@ -1571,3 +1571,172 @@ bool CTakeDownTicket::FindTicketInDb(const std::string& key, CTakeDownTicket& ti
+        }
+    } else {
+        if (fee != masterNodeCtrl.MasternodeUsernameChangeAgainFee) {
+            throw std::runtime_error(strprintf("%s ticket's fee is invalid. PastelID - [%s], invalid fee - [%" PRId64 "], expected fee - [%" PRId64 "]",
+                                               GetTicketDescription(TicketID::Username), pastelID, fee, masterNodeCtrl.MasternodeUsernameChangeAgainFee));
+        }
+    }
+    
+    return true;
+}
+
+CChangeUsernameTicket CChangeUsernameTicket::Create(std::string _pastelID, std::string _username, const SecureString& strKeyPass)
+{
+    CChangeUsernameTicket ticket(std::move(_pastelID), std::move(_username));
+
+    // Check if PastelID already have a username on the blockchain. 
+    if (!masterNodeCtrl.masternodeTickets.FindTicketBySecondaryKey(ticket)) {
+        // IF PastelID has no Username yet, the fee is 100 PSL
+        ticket.fee = masterNodeCtrl.MasternodeUsernameFirstChangeFee;
+    } else {
+        // IF PastelID changed Username before, fee should be 5000
+        ticket.fee = masterNodeCtrl.MasternodeUsernameChangeAgainFee;
+    }
+
+    ticket.GenerateTimestamp();
+    
+    std::string strTicket = ticket.ToStr();
+    ticket.signature = CPastelID::Sign(reinterpret_cast<const unsigned char*>(strTicket.c_str()), strTicket.size(), ticket.pastelID, strKeyPass);
+
+    return ticket;
+}
+
+bool CChangeUsernameTicket::FindTicketInDb(const std::string& key, CChangeUsernameTicket& ticket)
+{
+    ticket.username = key;
+    return masterNodeCtrl.masternodeTickets.FindTicket(ticket);
+}
+
+bool CChangeUsernameTicket::isUsernameBad(const std::string& username, std::string& error)
+{
+    std::string upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+    std::string lower = "abcdefghijklmnopqrstuvwxyz";
+    std::string digits = "0123456789";
+
+    // Check if has only <4, or has more than 12 characters
+    if ((username.size() < 4) || (username.size() > 12)) {
+        error = "Invalid size of username, the size should have at least 4 characters, and at most 12 characters";
+        return true;
+    }
+
+    // Check if doesn't start with letters.
+    if ( (upper.find(username.front()) == std::string::npos) && (lower.find(username.front()) == std::string::npos ) ) {
+        error = "Invalid username, should start with a letter A-Z or a-z only";
+        return true;
+    }
+    // Check if contains characters that is different than upper and lowercase Latin characters and numbers
+    if (!std::all_of(username.begin(), username.end(), [&](unsigned char c) 
+        { 
+          return (upper.find(c) != std::string::npos || lower.find(c) != std::string::npos || digits.find(c) != std::string::npos); 
+        }
+      )) {
+        error = "Invalid username, should contains letters A-Z a-z, or digits 0-9 only";
+        return true;
+    }
+    // Check if contains bad words (swear, racist,...)
+    for (const auto& elem : UsernameBadWords::Singleton().wordSet) {
```

Comment:
what about lowercase/upppercase/mixed case words? 
Make a copy of username and use lowercase() function from str_utils.h. Then you can search words from the badword set - they should be all lowercased as well.

Reply:
fixed